src/LtiServiceConnector.php

```diff
@@ -13,90 +15,99 @@ class LtiServiceConnector implements LtiServiceConnectorInterface
+    public function __construct(LtiRegistrationInterface $registration, Cache $cache, Client $client)
     {
         $this->registration = $registration;
+        $this->cache = $cache;
+        $this->client = $client;
     }
 
     public function getAccessToken(array $scopes)
     {
         // Don't fetch the same key more than once.
         sort($scopes);
-        $scope_key = md5(implode('|', $scopes));
-        if (isset($this->access_tokens[$scope_key])) {
-            return $this->access_tokens[$scope_key];
-        }
+        $scopeKey = md5(implode('|', $scopes));
 
         // Build up JWT to exchange for an auth token
-        $client_id = $this->registration->getClientId();
-        $jwt_claim = [
-                'iss' => $client_id,
-                'sub' => $client_id,
+        $clientId = $this->registration->getClientId();
+
+        // Store access token with a unique key
+        $accessTokenKey = $scopeKey.'-'.$clientId;
+
+        // Get Access Token from cache if it exists and is not expired.
+        if ($this->cache->getAccessToken($accessTokenKey)) {
+            return $this->cache->getAccessToken($accessTokenKey);
+        }
+
+        $jwtClaim = [
+                'iss' => $clientId,
+                'sub' => $clientId,
                 'aud' => $this->registration->getAuthServer(),
                 'iat' => time() - 5,
                 'exp' => time() + 60,
                 'jti' => 'lti-service-token'.hash('sha256', random_bytes(64)),
         ];
 
         // Sign the JWT with our private key (given by the platform on registration)
-        $jwt = JWT::encode($jwt_claim, $this->registration->getToolPrivateKey(), 'RS256', $this->registration->getKid());
+        $jwt = JWT::encode($jwtClaim, $this->registration->getToolPrivateKey(), 'RS256', $this->registration->getKid());
 
         // Build auth token request headers
-        $auth_request = [
+        $authRequest = [
             'grant_type' => 'client_credentials',
             'client_assertion_type' => 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer',
             'client_assertion' => $jwt,
             'scope' => implode(' ', $scopes),
         ];
 
-        // Make request to get auth token
-        $ch = curl_init();
-        curl_setopt($ch, CURLOPT_URL, $this->registration->getAuthTokenUrl());
-        curl_setopt($ch, CURLOPT_POST, 1);
-        curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query($auth_request));
-        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
-        curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
-        $resp = curl_exec($ch);
-        $token_data = json_decode($resp, true);
-        curl_close($ch);
-
-        return $this->access_tokens[$scope_key] = $token_data['access_token'];
+        $url = $this->registration->getAuthTokenUrl();
+
+        // Get Access
+        $response = $this->client->post($url, [
+            'timeout' => 10,
```

Comment:
Rather than setting this in the library, is there a way that we can configure it on the client that gets passed into it (and therefore, within our app instead of this library)?

Reply:
Would that work when timeout varies? For the auth token it's 10 (seconds?) but to make the service request timeout is 60. I'm assuming you mean to set a default timeout on the client?

Reply:
Hmm. Why is the auth token timeout 10 seconds, when the service request is 60? Where did you get those numbers from?

From my experience (and based on Canvas's rate limiting equation), the auth request is 5x slower than the service request on average (1.5s vs 0.3s).

I'd think we'd want to let whoever is using this request configure the timeout for the whole client, rather than forcing them to use the timeouts we decide. That does have the downside tho of not being able to vary the timeout between the auth and service request. My personal opinion is that it's fine to accept that downside, but I might not be understanding the full context.

Reply:
ah looks like the 10 & 60 second timeouts set in this file are arbitrary. Finally able to figure this out. so I removed the timeouts and created a client in GradebookSync13 and SendGradeBack with a timeout. I set it to 60 seconds.

https://github.com/packbackbooks/questions/pull/4455/files#diff-9e91242e398250a7a8bda45567746be7739b930668d09a87b10fc88562fee1d8R31-R34